woocommerce/woocommerce-gateway-stripe
# Add admin option to override subscription status during SEPA payment confirmation period

# Changes proposed in this Pull Request:

**Description:**

When SEPA is used to pay for a subscription (either parent or renewal) the subscription is set to an 'On Hold' status until payment confirmation is returned (can take up to 7-10 days). During this period customers do not have access to virtual or downloadable content, resulting in a shorter than desired period of access (a monthly subscription may only grant access for 1 month minus the 7-10 day waiting period).

This has primarily been handled by [a snippet](https://gist.github.com/jrick1229/99fff8a6b61580e5b0d6b000399573fe), but was requested to be added as a setting (due to the higher demand from Subscriptions/SEPA users) by customers and other HEs.

When this setting is activated - _Settings > Payments > Stripe SEPA > 'Subscriptions Status'_ - the subscription will be immediately set to 'Active', while the order is 'On Hold'. If the order is returned after payment confirmation to 'Failed', the subscription is switched to 'On Hold'. This same process is used for both parent and renewal orders, which both hit the `woocommerce_order_status_changed` action.

If this setting and function are not added then customers will have to go through support before being given the snippet to add, which we then don't support.

**Images or gifs:**

Subscription with setting not enabled:

<img width="1509" alt="Markup 2020-12-18 at 13 05 28" src="https://user-images.githubusercontent.com/19918755/102646333-d6548b80-4131-11eb-98fd-44277db84e6d.png">

With setting enabled: 

<img width="1510" alt="Screenshot on 2020-12-18 at 13-07-19" src="https://user-images.githubusercontent.com/19918755/102646433-069c2a00-4132-11eb-8f7c-1cbe7e10b9c6.png">

# Testing instructions

**Testing instructions:** How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?

Test with the setting enabled and disable, found in: _Settings > Payments > Stripe SEPA_

If setting is **disabled**, the subscription should be immediately set to 'On Hold', while the order is also 'On Hold'. To test a failure in this scenario, set the order status to 'Failed', the order should now be 'Failed' and the subscription should remain in an 'On Hold' status.
Test a renewal by setting the subscription's status to 'Active', then choose 'Process renewal' from the subscription dropdown. The subscription should follow the same status course as above, being set to 'On Hold' along with the order until payment is confirmed.

If setting is **enabled**, the subscription should be immediately set to 'Active', while the order is set to 'On Hold'. To test a failure in this scenario, set the order status to 'Failed', the order should now be 'Failed' and the subscription should transition to 'On Hold'.
Test a renewal by ensuring the subscription's status is 'Active', then choose 'Process renewal' from the subscription dropdown. The subscription should remain in an 'Active' state, while the order will be again set to 'On Hold' until payment is confirmed.

-------------------

(remove me)
